### PR TITLE
feat!: use numalogic v0.3

### PR DESCRIPTION
**Pull request by ab93**
Signed-off-by: Avik Basu <ab93@users.noreply.github.com>

### CODE REVIEW

File: manifests/pipeline/numalogic-prometheus-pipeline.yaml [hunk 1]
```diff
@@ -3,172 +3,193 @@ kind: Pipeline
 metadata:
   name: numalogic-prometheus-pipeline
 spec:
+  interStepBufferServiceName: numalogic-isbs
   watermark:
-    disabled: false
+    disabled: true
```
Comment (vigith) [inline manifests/pipeline/numalogic-prometheus-pipeline.yaml:8]: why disable?

Comment (ab93) [reply manifests/pipeline/numalogic-prometheus-pipeline.yaml]: we enabled it

File: manifests/pipeline/numalogic-prometheus-pipeline.yaml [hunk 2]
```diff
@@ -6,169 +6,196 @@ spec:
...
+            - udf
+            - metric_filter
+    - name: window
+      scale:
+        min: 1
+        max: 1
+      udf:
+        container:
+          image: quay.io/numaio/numalogic-prometheus/udf:v0.1.0a0
+          imagePullPolicy: Always
```
Comment (nkoppisetty) [inline manifests/pipeline/numalogic-prometheus-pipeline.yaml:43]: remove imagepullpolicy

File: manifests/prerequisites/numaflow/kustomization.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
-namespace: argo
+namespace: numaflow-system
 resources:
   - namespace.yaml
-  - https://github.com/argoproj/argo-workflows/releases/download/v3.3.10/install.yaml
+  - https://raw.githubusercontent.com/numaproj/numaflow/v0.6.4/config/install.yaml
```
Comment (nkoppisetty) [inline manifests/prerequisites/numaflow/kustomization.yaml:6]: v0.6.5

File: numaprom/udf/postprocess.py [hunk 4]
```diff
@@ -16,115 +19,125 @@
...
-    return labels
+def __construct_publisher_payload(
+    stream_payload: StreamPayload, final_score: float
+) -> PrometheusPayload:
+    metric_name = stream_payload.composite_keys["name"]
+    namespace = stream_payload.composite_keys["namespace"]
 
+    labels = {
+        "model_version": str(stream_payload.get_metadata("version")),
+        **stream_payload.get_metadata("src_labels"),
```
Comment (nkoppisetty) [inline numaprom/udf/postprocess.py:61]: do not add all the `src_labels`, add only what is required as it could lead to high cardinality data. And, especially in unified anomaly, it could take src_labels from a different metric each time, and will be regarded as a new time series. 

> CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

File: .dockerignore [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+manifests/
+docs/
+.github
+CHANGELOG.md
+CODE_OF_CONDUCT.md
+CONTRIBUTING.md
+README.md
+LICENSE
+.gitignore
+.idea/
```
Comment (nkoppisetty) [inline .dockerignore:11]: new line

File: .gitignore [hunk 6]
```diff
@@ -130,3 +130,6 @@ dmypy.json
 
 # Jetbrains project
 .idea/
+
+# Mac related
+.DS_Store
```
Comment (nkoppisetty) [inline .gitignore:135]: new line

File: .dockerignore (hunk 5 above)
Comment (nkoppisetty) [reply .dockerignore:11]: add .codecov, .coveragerc, .flake8 too?

File: numaprom/entities.py [hunk 7]
```diff
@@ -14,54 +19,45 @@ class Status(str, Enum):
...
+class StreamPayload:
     uuid: str
-    metric_name: str
-    key_map: Dict
-    src_labels: Dict[str, str]
-    processedMetrics: List[Metric]
-    startTS: str
-    endTS: str
+    win_arr: Matrix
+    win_ts_arr: List[str]
```
Comment (nkoppisetty) [inline numaprom/entities.py:32]: what is the difference between win_arr and win_ts_arr?

File: numaprom/udf/inference.py [hunk 8]
```diff
@@ -21,75 +20,60 @@
...
-        return "", ""
-    return "postprocess", payload.to_json()
+def _run_model(
+    payload: StreamPayload, artifact_data: ArtifactData, model_config: Dict
+) -> Tuple[str, str]:
+    model = artifact_data.artifact
+    stream_data = payload.get_streamarray()
+    streamloader = DataLoader(StreamingDataset(stream_data, model_config["win_size"]))
+
+    trainer = AutoencoderTrainer()
```
Comment (nkoppisetty) [inline numaprom/udf/inference.py:30]: We might need to take this too from config, to make it work with different kinds of models. We can do it later in the future.

File: numaprom/udf/postprocess.py (hunk 4 above)
Comment (nkoppisetty) [reply numaprom/udf/postprocess.py:61]: Add hash_id as a label for sure

File: pyproject.toml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "numalogic-prometheus"
-version = "0.0.3"
+version = "0.1.0a0"
```
Comment (nkoppisetty) [inline pyproject.toml:3]: version 0.0.4? we can do 0.1.0 after trainer with redis?

File: numaprom/udf/inference.py (hunk 8 above)
Comment (ab93) [reply numaprom/udf/inference.py:30]: Yes will do it later, when we tackle config management

File: numaprom/entities.py (hunk 7 above)
Comment (ab93) [reply numaprom/entities.py:32]: `win_arr` is only for the actual time series. Can take in multivariate too.
`win_ts_arr` is the timestamp array

File: numaprom/udf/postprocess.py (hunk 4 above)
Comment (ab93) [reply numaprom/udf/postprocess.py:61]: Thanks for this. Can we define, what all labels do we need? It will help us to create a structure, and not be surprised :)

File: numaprom/entities.py (hunk 7 above)
Comment (nkoppisetty) [reply numaprom/entities.py:32]: we can't have them in one?

File: numaprom/udf/postprocess.py (hunk 4 above)
Comment (nkoppisetty) [reply numaprom/udf/postprocess.py:61]: Ideally, it should just be the ones from composite keys excluding the metric name. So, just namespace for argo cd anomaly, namespace and hash_id for rollouts anomaly. 

If the users want some extra labels to be added, we could add it to metric config. (in the future)

**Review by nkoppisetty — APPROVED** [review]
LGTM and great work!! :)
